spring-integration-core/src/main/java/org/springframework/integration/util/SimplePool.java

```diff
@@ -102,11 +102,9 @@ public synchronized void setPoolSize(int poolSize) {
 					break;
 				}
 				T item = this.available.poll();
-				if (item == null) {
-					this.permits.release();
```

Comment:
You have stopped to release permits.
Why is that, please?

Reply:
In this branch we need  to call only `acquire()`. Original intent was to revert acquire if item==null. But this is absolutely incorrect. For example, if we have a new instance of pool, then "available" collection is empty. But setPoolSize() method should correctly update all structures. After calling `setPoolSize(5)` everything should contain `5`. (permits==5, poolSize==5, and targetPoolSize==5). Of course, I am describing the situation when no objects were allocated. Please, check added test cases.

Reply:
I have added additional checks to test that "permits" field is correctly updated